File: go/src/book/GoDS/7_sort/quickSort.go
package main

func quickSort(values []int, left, right int) {
	temp := values[left]
	p := left
	i, j := left, right

	for i <= j {
		//找到第一个比temp小的数
		for j >= p && values[j] >= temp {
			j--
		}
		//将该数值赋值给p
		if j >= p {
			values[p] = values[j]
			p = j
		}
		//找到第一个比temp大的数
		if values[i] <= temp && i <= p {
			i++
		}
		//将该数值赋值给p
		if i <= p {
			values[p] = values[i]
			p = i
		}
	}
	values[p] = temp
	if p-left > 1 {
		quickSort(values, left, p-1)
	}
	if right-p > 1 {
		quickSort(values, p+1, right)
	}

}

func QuickSort(values []int) {
	quickSort(values, 0, len(values)-1)
}

File: code_book/TEMPLATE/c/sort/quicksort.c
#include <stdio.h>
#include "../common_header/utils.h"

void quicksort(int v[], int left, int right) {
    int i, last;
    void swap(int v[], int i, int j);
    if (left >= right)
        return;
    swap(v, left, (left + right)/2);
    last = left; 
    for (i = left + 1; i <= right; i++)
        if (v[i] < v[left])
            swap(v, ++last, i);
    swap(v, left, last);
    quicksort(v, left, last - 1);
    quicksort(v, last + 1, right);
}

void swap(int v[], int i, int j) {
    int temp;
    temp = v[i];
    v[i] = v[j];
    v[j] = temp;
}

int main() {
    int a[] = {5,2,3,8,10,21,32,55,14};
    int l = sizeof(a)/sizeof(a[0]);
    quicksort(a, 0, l);
    print_array(a);
}